File: src/pythonic/benchmark.py
import time
import random
from balance_multiply import BalanceMultiply
from naive_multiply import Naive
from karatsuba_multiply import Karatsuba
from power_of_two_multiply import MutiplyPowerTwo

import sys
sys.set_int_max_str_digits(10000)


def benchmark():
    def bulitin(a, b): return a * b
    naive = Naive()
    karatsuba = Karatsuba()
    power = MutiplyPowerTwo()
    balance = BalanceMultiply()

    methods = {
        "Python Built-in": bulitin,
        "Naive": naive.multiply,
        "Karatsuba": karatsuba.multiply,
        "PowerTwo": power.multiply,
        "Balance": balance.multiply
    }

    input_sizes = [x for x in range(1, 100)]
    runs = 50
    results = {method: [] for method in methods}

    for size in input_sizes:
        numA = 1024
        numB = random.randint(10**(size - 1), 10**size - 1)

        print(f"\nTesting with input size: {size} digits")
        for name, func in methods.items():
            total_time = 0
            for _ in range(runs):
                start_time = time.time_ns()
                func(numA, numB)  # Run the multiplication
                total_time += (time.time_ns() - start_time)

            avg_time = total_time / runs
            results[name].append(avg_time)
            print(f"{name} Avg Time: {avg_time / 1e9:.6f}s")

    return results


benchmark_results = benchmark()

File: src/benchmark.rs
extern crate num_bigint;
extern crate rand;

use num_bigint::BigUint;
use rand::Rng;
use std::time::Instant;

use crate::karatsuba_multiply::MultiplyKaratsuba;
use crate::naive_multiply::NaiveMultiplication;
use crate::power_multiply::MultiplyPowerTwo;

pub fn benchmark() -> std::collections::HashMap<String, Vec<u128>> {
    let mult = MultiplyPowerTwo::new();

    let methods: Vec<(&str, Box<dyn Fn(&BigUint, &BigUint) -> BigUint>)> = vec![
        ("Naive", Box::new(|a, b| NaiveMultiplication.multiply(a, b))),
        (
            "Karatsuba",
            Box::new(|a, b| MultiplyKaratsuba::multiply(a, b)),
        ),
        ("PowerTwo", Box::new(|a, b| mult.multiply(a, b))),
    ];

    let input_sizes = 1..=10;
    let runs = 50;
    let mut results: std::collections::HashMap<String, Vec<u128>> =
        std::collections::HashMap::new();

    for size in input_sizes {
        let num_a = BigUint::from(1024u64);
        let num_b = BigUint::from(rand::thread_rng().gen_range(0..10u64.pow(size) - 1));

        println!("\nTesting with input size: {} digits", size);

        for (name, func) in methods.iter() {
            let mut total_time = 0;
            for _ in 0..runs {
                let start_time = Instant::now();
                func(&num_a, &num_b);
                total_time += start_time.elapsed().as_nanos();
            }

            let avg_time = total_time / runs as u128;
            results
                .entry(name.to_string())
                .or_insert_with(Vec::new)
                .push(avg_time);
            println!("{} Avg Time: {}ns", name, avg_time);
        }
    }

    results
}
